Pprof modularization

pippolo84 opened the pull request:
Modularize the pprof related code into a cell. The cell is used by the agent, the operator and the clustermesh-apiserver to register the pprof HTTP handlers and export runtime profiling data.
    
The already present `pprof.Enable` function is marked as deprecated, but not removed, since hubble-relay still depends on it and that is not yet modularized to use hive and cells.
    
Differently from the agent and the clustermesh-apiserver, the operator prefixes the pprof related flags with `operator-`.  To reuse the same cell, we use an operator specific config type, in order to map the equivalent fields to the operator-specific flag names.  Also, when importing the cell, an additional call to cell.Config is requested to register the default config type (for agent and clustermesh-apiserver) or the operator-specific one.

The PR also adds an integration-style unit test to verify the correct registration of the pprof HTTP handlers when the cell is used.

Fixes: #24048

pippolo84 committed:
pprof: Add a cell for pprof HTTP handlers

Modularize pprof related code into a cell. The cell is used by the
agent, the operator and the clustermesh-apiserver.

The already present pprof.Enable function is marked as deprecated.

Differently from the agent and the clustermesh-apiserver, the operator
prefixes the pprof related flags with `operator-`.  To reuse the same
cell, we use an operator specific config type, in order to map the
equivalent fields to the operator-specific flag names.  Also, when
importing the cell, an additional call to cell.Config is requested to
register the default config type (for agent and clustermesh-apiserver)
or the operator-specific one.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
pprof: Add a test for the pprof cell

Add two unit-style integration tests to verify the correct registration
of the pprof HTTP handlers. Specifically, the first one check that no
listeners is started when pprof is disabled and the other one exercises
the `debug/pprof/heap` endpoint when pprof is enabled.  Both tests
verify that the cell is not leaking any goroutines after the hive stops.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/pprof/cell_test.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
```
Realized yesterday that you can also do: 
```
func TestMyTest(t *testing.T) {
    defer goleak.VerifyNone(t)
   ...
}
```

Probably nicer in cases like this?

joamaki commented on pkg/pprof/cell.go:
```diff
@@ -0,0 +1,82 @@
+			log.Info("Started pprof server")
+			go func() {
+				if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+					log.WithError(err).Error("server stopped unexpectedly")
```
Wonder if we should call `Shutdowner.Shutdown()` here. Now we'd just log an error if we fail to bind to pprof port. I guess it isn't really a fatal error...

pippolo84 replied:
Exactly, that was my thought.
I see that for [metrics server](https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L1868) we call `Shutdowner.Shutdown`, but I think that for pprof it is reasonable to leave out the error, apart from logging it.

squeed replied:
agreed, since these are host-network pods, the chance of a port collision is non-zero. And it really sucks to lose networking because of that. Better to log and continue.

joamaki commented on operator/cmd/flags.go:
```diff
@@ -246,12 +247,6 @@ func init() {
-
-	flags.Int(operatorOption.PProfPort, defaults.PprofPortOperator, "Port that the pprof listens on")
-	option.BindEnv(Vp, operatorOption.PProfPort)
-
```
there was no `operator-pprof-address`?

pippolo84 replied:
Nope, it was missing.

joamaki commented on operator/option/config.go:
```diff
@@ -94,13 +94,13 @@ const (
 	OperatorPrometheusServeAddr = "operator-prometheus-serve-addr"
 
 	// PProf enabled pprof debugging endpoint
-	PProf = "operator-pprof"
```
Perhaps we could just remove these?

joamaki commented on pkg/defaults/defaults.go:
```diff
@@ -20,13 +20,16 @@ const (
 	PprofPortAgent = 6060
 
-	// PprofPortAgent is the default value for pprof in the operator
+	// PprofPortOperator is the default value for pprof in the operator
```
Remove and just do the defaults inline where we define the default config? Makes more sense to me that way, but happy to hear if someone has an argument for keeping `pkg/defaults`.

pippolo84 replied:
Unfortunately those defaults are used also by [bugtool](https://github.com/cilium/cilium/blob/master/bugtool/cmd/root.go#L90-L95). Considered that, it probably makes sense to leave them in `pkg/defaults`.

pippolo84 replied:
As a trade-off, I ended up moving the default values from `pkg/defaults` to the binary specific option package (`pkg/option` for the agent, `operator/option` for the operator and so on).

joamaki commented on pkg/pprof/cell_test.go:
```diff
@@ -0,0 +1,60 @@
+		Cell,
+		cell.Config(Config{
+			Pprof:     true,
+			PprofPort: 8080,
```
This is a really often used port so this is likely to generate a lot of flakes. Preferably we'd allow the kernel to allocate the port, e.g. by using a custom `net.Listener` and then querying what it got via `Addr()`. Or at the very least let's pick a more uncommon port.

YutaroHayakawa reviewed (approved):
ClusterMesh part looks good. Thanks!

pippolo84 committed:
docs: Regenerate cmdref following pprof modularization

This commit regenerates the cmdref documentation files following the
modularization of pprof.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
hive: Fix typo in ProvidePrivate documentation

Fix a typo (`Private` -> `Provide`) in ProvidePrivate function inline
documentation.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

squeed commented on operator/cmd/flags.go:
```diff
@@ -330,3 +325,20 @@ func init() {
+	OperatorPprofPort    int
+}
+
+func (def operatorPprofConfig) Flags(flags *pflag.FlagSet) {
```
Should this be a pointer receiver?

pippolo84 replied:
Since there's no need to modify the receiver and copying an `operatorPprofConfig` instance is very cheap (besides, this method should be called just once at startup), I think it is fine to use a value receiver.

squeed replied:
sounds good.

joamaki commented on pkg/pprof/cell.go:
```diff
@@ -0,0 +1,126 @@
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool(Pprof, def.Pprof, "Enable serving pprof debugging API")
+	flags.String(PprofAddress, def.PprofAddress, "Address that pprof listens on")
+	flags.Int(PprofPort, def.PprofPort, "Port that pprof listens on")
```
Can we use uint16 here instead?

joamaki commented on pkg/pprof/cell.go:
```diff
@@ -0,0 +1,126 @@
+// profiling data in the format expected by the pprof visualization tool.
+var Cell = cell.Module(
+	"pprof",
+	"pprof HTTP handlers",
```
I find talking about "HTTP handlers" here (and in cells.go) confusing as we're actually running an HTTP server and not providing handlers to some other HTTP server. I would instead title this as: "Exposes pprof runtime profiling over HTTP" or something like that.

joamaki commented on pkg/pprof/cell.go:
```diff
@@ -0,0 +1,126 @@
+	}
+	s.listener = listener
+
+	log = log.WithFields(logrus.Fields{
```
Use `s.logger` instead

joamaki commented on pkg/pprof/pprof.go:
```diff
@@ -17,6 +17,8 @@ import (
 var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "pprof")
```
Remove and use injected `logrus.FieldLogger`

pippolo84 replied:
This is used by `pprof.Enable`, that we still need for hubble-relay. Being outside of the cell lifecycle, we won't have any injected `logrus.FieldLogger` here.
But this should be gone once hubble-relay will be refactored to use the pprof cell.

joamaki commented on pkg/pprof/cell.go:
```diff
@@ -0,0 +1,126 @@
+		"port": s.listener.Addr().(*net.TCPAddr).Port,
+	})
+
+	s.httpSrv = &http.Server{}
```
I'm slightly worried about the use of the default mux here. Consider instead building a new mux here with just the handlers for pprof instead so we'll know exactly what we're serving and not potentially random stuff added by imported packages.

Check pprof.init for what the handlers need to be: https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/net/http/pprof/pprof.go;l=85;bpv=1

pippolo84 committed:
options: Move pprof defaults into binary specific packages

Move pprof related options default values (address and port) from
pkg/defaults to binary specific packages, like pkg/option for the agent,
operator/option for the operator and clustermesh-apiserver/option for
the clustermesh server.
This is done to avoid polluting the global defaults package with
defaults that are not really global but binary specific.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
